numba/cuda/simulator/cudadrv/devicearray.py

```diff
@@ -30,23 +30,59 @@ class FakeShape(tuple):
     indexing, similar to the shape in CUDA Python. (Numpy shape arrays allow
     negative indexing)
     '''
+
     def __getitem__(self, k):
         if isinstance(k, int) and k < 0:
             raise IndexError('tuple index out of range')
         return super(FakeShape, self).__getitem__(k)
 
 
+class FakeWithinKernelCUDAArray(object):
+    '''
+    Created to emulate the behavior of arrays within kernels, where either array.item or array['item'] is valid.
```

Comment:
As far as I can tell from the code, this is to emulate the behaviour of structured types or records in kernels - if an array is passed in to `FakeCUDAArray.__init__`, then a `FakeWithinKernelCUDAArray` is not created, but the original array is used for its `_ary_access` attribute.

Am I misunderstanding what's happening here?

Reply:
I think so. Note that a FakeCUDAArray becomes a FakeWithinKernelCUDAArray in a FakeCUDAKernel. This emulates the behavior that you can only say `array.item` or `array['item']` within a kernel, and not outside.

Reply:
This line is too long, please wrap to 80 chars.